library/std/src/sys/random/linux.rs

```diff
@@ -0,0 +1,153 @@
+//! for the output of `DefaultRandomSource` so we simply wait until it is
+//! initialized. When `HashMap` keys however, this represents a potential source
+//! of deadlocks, as the additional entropy may only be generated once the
+//! program makes forward progress. In that case, we just use the best random
+//! data the system has available at the time.
+//!
+//! So in conclusion, we always want the output of the non-blocking pool, but
+//! may need to wait until it is initalized. The default behaviour of `getrandom`
+//! is to wait until the non-blocking pool is initialized and then draw from there,
+//! so if `getrandom` is available, we use its default to generate the bytes. For
+//! the best-effort data, however, we need to specify the `GRND_INSECURE` flags,
+//! but that is only available starting with kernel version 5.6. Thus, if we
+//! detect that the flag is unsupported, we try `GRND_NONBLOCK` instead, which
+//! will only succeed if the pool is initialized. If it isn't, we fall back to
+//! the file access method.
+//!
+//! The behaviour of `/dev/urandom` is inverse to that of `getrandom`: it always
+//! yields data, even when the pool is not initialized. For generating `HashMap`
+//! keys, this is not important, so when we only need best-effort data, we can
+//! use it directly. For secure data however, we need to wait until
+//! initialization, which we can do by `poll`ing `/dev/random`.
+//!
+//! TLDR: our fallback strategies are:
+//!
+//! Secure data                                 | Best-effort data
+//! --------------------------------------------|------------------
+//! getrandom(0)                                | getrandom(GRND_INSECURE)
+//! poll("/dev/random") && read("/dev/urandom") | getrandom(GRND_NONBLOCK)
+//!                                             | read("/dev/urandom")
+//!
+//! [^1]: <https://lwn.net/Articles/606141/>
+//! [^2]: <https://lwn.net/Articles/808575/>
+//!
+// FIXME(in 2040 or so): once the minimum kernel version is 5.6, remove the
+// `GRND_NONBLOCK` fallback and use `/dev/random` instead of `/dev/urandom`
+// when secure data is required.
+
+use crate::fs::File;
+use crate::io::Read;
+use crate::os::fd::AsRawFd;
+use crate::sync::atomic::AtomicBool;
+use crate::sync::atomic::Ordering::Relaxed;
+use crate::sync::OnceLock;
+use crate::sys::pal::os::errno;
+use crate::sys::pal::weak::syscall;
+
+// A weak symbol allows interposition, e.g. for perf measurements that want to
+// disable randomness for consistency. Otherwise, we'll try a raw syscall.
+// (`getrandom` was added in glibc 2.25, musl 1.1.20, android API level 28)
+syscall! {
+    fn getrandom(
+        buffer: *mut libc::c_void,
+        length: libc::size_t,
+        flags: libc::c_uint
+    ) -> libc::ssize_t
+}
+
+static GETRANDOM_AVAILABLE: AtomicBool = AtomicBool::new(true);
+static GRND_INSECURE_AVAILABLE: AtomicBool = AtomicBool::new(true);
+static URANDOM_READY: AtomicBool = AtomicBool::new(false);
```

Comment:
Is the extra boolean needed? `OnceLock` can have the same atomic check done by calling `.get()` (which doesn't block) and avoid needing to store readiness in multiple places.

Reply:
No, that's not correct since `/dev/urandom` is also used for insecure random data, which would initialize it before the `poll` would return.